### DISCUSSION TITLE
Subclass a rust class from python

### DISCUSSION THREAD

**Discussion by mirage2032** [PyO3/pyo3#3938] [Questions]
I'll try to shortly explain what I'm trying to do
In rust I want to have a class A with a foo method that calls bar
```rust

#[pymethods]
impl PyMemDevice {
    #[new]
    fn new(_args: Option<&PyTuple>, _kwargs: Option<&PyDict>) -> Self {
        Python::with_gil(|py|
            PyMemDevice {
            }
        )
    }
    
    fn foo(&self) -> PyResult<()> {
        self.bar();
        Ok(())
    }
}
```

In python I want to implement the B class that inherits from A and implements bar()
Similar to this
```python
class B(A):
    def bar():
        pass
```

How would I make rust know that bar() is an abstract method and to call it's python implementation?
I did quite a few experiments on this but with no success

**Comment by davidhewitt** [comment]
You should write `foo` like this:

```rust
    fn foo(slf: &Bound<Self>) -> PyResult<()> {
        self.call_method0("bar")?;
        Ok(())
    }
```

NB this uses the `Bound` API only available in 0.21 (already released as a beta version).

**Reply by mirage2032** [reply]
Will check it out, thanks.

**Reply by toppk** [reply]
this is cool, been looking for this for a while.  it it possible to do a subclasss check so if the class isn't subclassed, that you just call the base implemention in rust without a roundtrip though the gil?  Or does call_method0 do that check itself before taking the gil?

**Reply by birkenfeld** [reply]
No, it doesn't do such checks; `call_method0` is implemented for `PyAny` and doesn't know anymore you're calling it from a Rust pyclass.

But yes, you can do the check yourself using `self.is_exact_instance_of::<PyMemDevice>()`.

**Reply by toppk** [reply]
@birkenfeld Thanks.   That was the approach I was figuring on.  But then I have slf a Bound, and not self, my object with access to all my functions that I could just directly call them.   Is there a way to get the "real" self?   (i've been reading through github search results and it seems like no one really provides classes that people are able to override bits and pieces of it like this, but i'm surprised I didn't see anyone using self and slf at the same time somehow).  My guess is I cannot use bound, but have to go via py, but i'm still so new to rust to figure this out on my own.

**Reply by birkenfeld** [reply]
You should be able to call this on `slf` as well.

**Reply by toppk** [reply]
What I'm trying to do is the following:
```rust
#[pymethods]
impl MyClass {
  fn a_method(&self) {
      println!("base a_method");
  }
  fn foo(slf: &Bound<Self>) -> PyResult<()> {
      if slf.is_exact_instance_of::<MyClass>() {
          slf.a_method();
      } else {
          slf.call_method0("a_method")?;
      }
      Ok(())
  }
}
```
But slf, isn't MyClass or I don't know how to get to my instance from slf.

**Reply by toppk** [reply]
```rust
            let mut this: PyRefMut<Self> = slf.try_borrow_mut()?;
            let this: &mut Self = &mut *this;
```

I found this incantation, that seems to allow me to call my methods directly, but this move's slf, so I'm not sure if that's the right approach.  I'm not even clear I'll need all this, but I guess I tried to replicate the way I do things in python in rust (e.g. having base classes that call a lot of other methods that may or may not be implemented in the base class).  I think it's time for me to think about this more in a composition way rather than inheritance way.
